File: python/rail_fence_cipher.py
def encode(message, rails):
    fences = [""] * rails
    i = 0
    sign = 1
    for c in message:
        fences[i] += c
        val = i + sign 
        if val == -1 or val == rails:
            sign *= -1
        i += sign

    return "".join(fences)


def decode(encoded_message, rails):
    dist = 2 * rails - 2
    l = len(encoded_message)
    k = 0
    decode_message = [""] * l
    for i in range(rails):
        offset = dist - 2 * i 
        j = i
        for j in range(i, l, dist):
            if k < l: 
                decode_message[j] = encoded_message[k]
                k += 1
            if i != 0 and i != rails-1 and j+offset < l:
                decode_message[j+offset] = encoded_message[k]
                k += 1
        
    return "".join(decode_message)

File: go/rail_fence_cipher.go
package railfence

import "strings"

func Encode(message string, rails int) string {
	fences := make([]string, rails)
	i, sign := 0, 1

	for _, c := range message {
		fences[i] += string(c)
		if val := i + sign; val == -1 || val == rails {
			sign *= -1
		}
		i += sign
	}

	return strings.Join(fences, "")
}

func Decode(message string, rails int) string {
	dist := 2*rails - 2
	k, l := 0, len(message)
	decode_message := make([]byte, l)

	for i := 0; i < rails; i++ {
		offset := dist - 2*i // offset of next char
		for j := i; j < l; j += dist {
			if k < l { // add a new letter every "post"
				decode_message[j] = message[k]
				k++
			}
			if i != 0 && i != rails-1 && j+offset < l { // add a second letter if on diag
				decode_message[j+offset] = message[k]
				k++
			}
		}

	}
	return string(decode_message)
}
